### PR TITLE
Fix maintainVisibleContentPosition in Safari

### PR DESCRIPTION
**Pull request by janicduplessis**
### Details

The maintainVisibleContentPosition implementation currently doesn't work properly. This fixes it.

I noticed some cases that caused some jumping:

- Sometimes the first visible view will be removed from the DOM and it would cause an incorrect adjustment. To fix this we can detect in the mutation observer that the node removed is the first visible view, and calculate a new one when it happens. Also reduce the pagination size to reduce cases where the first visible view could be removed.

- We do the adjustment in a `requestAnimationFrame`, but this gives time for a scroll event to happen before we get to adjust the scroll position and we lose the proper positioning. It seemed to be needed in chrome, but I found a better way, we can instead use the last scroll position from the previous scroll event instead and it works fine without the need of `requestAnimationFrame` on both chrome and safari.

- Add some hacks to mobile safari since adjusting scroll position during momentum scroll does not work.

Before:

https://github.com/user-attachments/assets/13eaa301-f557-4c14-be7c-015fc539443e

After:

https://github.com/user-attachments/assets/d39fb102-2211-45e6-91ea-a1fc001684b1

### Fixed Issues

$ https://github.com/Expensify/App/issues/47518
PROPOSAL:

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] Check that scrolling down is smooth after comment linking using the following steps

- In a chat with many messages copy the link to a message and send it
- Click on the message link from the first step
- Scroll down

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

N/A

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/9eacba9d-0024-4c91-9812-8fb7b11f244d

</details>

<details>
<summary>iOS: Native</summary>

N/A

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/f08d83f0-41fe-4547-a93c-fe4fe871175d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Safari

https://github.com/user-attachments/assets/937c653f-8a7f-4512-969e-d0a2caac57a4

Chrome

https://github.com/user-attachments/assets/4ac0050d-1e9f-4328-a9bd-1276179d5114

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/0df58b82-9be3-42cc-a720-303ac1224703

</details>

### CODE REVIEW

**Comment by janicduplessis** [comment]
After doing more testing it seems like Safari on iOS is still problematic, this only fixes desktop. Investigating.

**Comment by janicduplessis** [comment]
The issue on mobile safari seems to be very hard to fix, I have tried many things but I haven't been able to get safari to change the scroll position while a momentum scroll is happening.

The only workaround I found is to disable scrolling then update the scroll position then re-enable scrolling, but this will cause the momentum scroll to stop which makes scrolling not smooth, but in my opinion a pretty good improvement over a giant jump like it does currently.

I will try to investigate some other websites that have bidirectional scroll to see if they handle mobile safari better than this.

**Comment by janicduplessis** [comment]
I tested slack on iOS safari and they seem to use a similar workaround, the scroll down after linking to a message is a bit jumpy and momentum scroll keeps stopping.

**Comment by janicduplessis** [comment]
@ishpaul777 This is ready if you wanna have a look!

**Comment by janicduplessis** [comment]
@ishpaul777 Can you try again to see if the scroll still gets stuck sometimes, I just merged main which has the fixes from https://github.com/Expensify/App/pull/46315.

**Comment by janicduplessis** [comment]
I cannot currently reproduce the blanking issue, but I did find one more problem that I would like to address before merging this.

If I link to the first message, then scroll back all the way to the bottom of the chat back to where the links are, then link to the first message again the scroll position is wrong.

**Review by ishpaul777 — APPROVED** [review]
Looks good! just one comment

File: src/CONST.ts [hunk 1]
```diff
@@ -4483,7 +4483,7 @@ const CONST = {
     REPORT_FIELD_TITLE_FIELD_ID: 'text_title',
 
     MOBILE_PAGINATION_SIZE: 15,
-    WEB_PAGINATION_SIZE: 50,
+    WEB_PAGINATION_SIZE: 30,
```
[the remark this answers is not in the document]
Comment (janicduplessis) [reply src/CONST.ts:4486]: This seemed to help reduce jumping, when loading more messages I noticed that often the first visible view would be removed. When reducing this it stopped happening.

**Review by ishpaul777 — APPROVED** [review]

**Comment by stitesExpensify** [comment]
@janicduplessis would you mind taking one more look here as well?

**Comment by janicduplessis** [comment]
@stitesExpensify Sorry I am not sure what you mean here

**Comment by stitesExpensify** [comment]
@janicduplessis so sorry, I didn't have my head on straight this morning. I thought that you were a reviewer not the author 🤦

**Review by stitesExpensify — APPROVED** [review]

**Pull request merged**